### PR TITLE
Converted test to docbook test and made the test more readable

### PR DESCRIPTION
**Pull request by johlrogge**
I thought resiter would benefit from examples in the api-docuentation. I made this change to illustrate.

When I looked at the `and_then_filter.rs` it looks lke it impleents `try_filter_map_ok` as per the previous pullrequest naming. Perhaps `try_filter_ok`whould move from `filter`to this module? And this module could be renamed to `try_filter`to collect all trelated try_filter-methods?

If that sounds like a good idea I would be happy to put in some work to get it done, and if you agree that doctests would be a good idea I'd be happy to put in some work to give the API-documenttaion some love by primarily converting some of the existing tests to doctests.

### CODE REVIEW

File: src/and_then_filter.rs [hunk 1]
```diff
@@ -6,6 +6,50 @@
 
 /// Extension trait for `Iterator<Item = Result<O, E>>` to selectively transform and map Oks and Errors.
 pub trait AndThenFilter<O, E>: Sized {
+    /// Equivalent to [Iterator::filter_map] on all `Ok` values.
+    /// The filter function can fail with a result and turn an
+    /// [Result::Ok] into a [Result::Err]
+    /// ```
```
Comment (matthiasbeyer) [inline src/and_then_filter.rs]: Can you add a newline between the text and the code example please?

File: src/and_then_filter.rs [hunk 2]
```diff
@@ -6,6 +6,50 @@
...
+    /// use std::str::FromStr;
+    /// use resiter::and_then_filter::AndThenFilter;
+    ///
+    /// let filter_mapped: Vec<_> = vec![
```
Comment (matthiasbeyer) [inline src/and_then_filter.rs:17]: I am not entirely sure here, but you might not need `vec![]` here, a normal slice `[...]` should be enough, isn't it?

File: src/and_then_filter.rs [hunk 3]
```diff
@@ -6,6 +6,50 @@
...
+    ///     Ok("a"), // will become an error
+    ///     Err("4".to_owned()),
+    ///     Ok("5"), // will be filtered out
+    ///     Err("b".to_owned()),
+    ///     Err("8".to_owned()),
+    /// ]
+    /// .into_iter()
+    /// .and_then_filter(|txt| {
+    ///     let r = usize::from_str(txt).map_err(|e| e.to_string());
+    ///     match r {
```
Comment (matthiasbeyer) [inline src/and_then_filter.rs]: I think it would be okay to collapse these two lines into one:

```suggestion
    ///     match usize::from_str(txt).map_err(|e| e.to_string()) {
```

File: src/and_then_filter.rs [hunk 4]
```diff
@@ -38,15 +82,14 @@ where
...
                     None => continue,
                 },
-                Some(Err(e)) => return Some(Err(e)),
-                None => return None,
-            }
+                Some(Err(e)) => Some(Err(e)),
+                None => None,
+            };
         }
     }
```
Comment (matthiasbeyer) [inline src/and_then_filter.rs:94]: I like what you did here, but could you split it to a separate commit please?

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]
Hi!

Thanks a lot for this patch! I generally like it, only some minor things... except the "separate patch", that's a bit more important to me!

:smile:

File: src/and_then_filter.rs (hunk 2 above)
Comment (johlrogge) [reply src/and_then_filter.rs:17]: I tried just  to remove the `vec!` but can't really decipher what the compiler tells me when I do that. I will leave this one for now and get back to it.

File: src/and_then_filter.rs (hunk 4 above)
Comment (johlrogge) [reply src/and_then_filter.rs:94]: I agree. Will make that a separate commit

**Review by matthiasbeyer — APPROVED** [review]

**Comment by matthiasbeyer** [comment]
Thanks for your contribution! :tada: 

bors merge
